### DISCUSSION TITLE
Defining negative feature patterns in AttributeRuler

### DISCUSSION THREAD

**Discussion by megamattc** [explosion/spaCy#11755] [Help: Coding & Implementations] [answered]
Hello,

I was wondering if it is possible to define a negative feature pattern in the AttributeRuler? By this, I mean the following. I have two lexical items written the same way but with different grammatical functions, and I want to assign different UPOS, XPOS labels to them:

ša1 = relative pronoun -> "attrs": "POS:PRON", "TAG":"PRON"
ša2 = preposition -> "attrs": "POS:ADP", "TAG":"PRP

In telling the AttributeRuler how to handle these two items I can specify the pattern for ša2 as "ORTH":"ša", "DEP":"case", because ša2 always takes case as its syntactic dependency. But for ša1, there can be many dependency types it takes (nsubj, obl, iobj, etc.), all except case. So I would like to specify a pattern where "DEP" is anything but case. Is there a way to do this, perhaps using the '!' symbol?

**Comment by polm** [comment, marked as answer]
Negative operators work as usual in AttributeRuler patterns. In this case you can use `NOT_IN`, like so:

```
{"DEP": {"NOT_IN": ["case"]}}
```

`!` also works as usual, but it negates a whole pattern, while you only want to negate part of a pattern here, since you want to match in a positive fashion on the token text.

**Comment by megamattc** [comment]
Ah, somehow I overlooked that operator in the API, and didn't see an example of it. Thanks.

**Comment by polybiusproxy** [comment]
Uhm, not related to this topic, but did anyone else get mentioned by a scam NSFW bot?

![image](https://user-images.githubusercontent.com/47796739/209932532-77058c90-6622-47e4-b907-c0e88226160c.png)

**Reply by polm** [reply]
There has been a flood of comments like that starting in the past 24h or so. They've been reported and are getting deleted quickly, but unfortunately you might see messages in mail notifications or elsewhere. 

We've taken steps to make sure it doesn't continue and are monitoring the situation.
